src/components/flowchart/index.js

```diff
@@ -328,18 +329,13 @@ export class FlowChart extends Component {
               <path d="M 0 0 L 10 5 L 0 10 L 4 5 z" />
             </marker>
           </defs>
-          <g ref={el => (this._gWrapper = el)}>
-            <g ref={el => (this._gInner = el)}>
-              <g
-                className="pipeline-flowchart__edges"
-                ref={el => (this._gEdges = el)}
-              />
-              <g
-                id="nodes"
-                className="pipeline-flowchart__nodes"
-                ref={el => (this._gNodes = el)}
-              />
-            </g>
+          <g ref={this.wrapperRef}>
+            <g className="pipeline-flowchart__edges" ref={this.edgesRef} />
+            <g
+              id="nodes"
+              className="pipeline-flowchart__nodes"
+              ref={this.nodesRef}
+            />
           </g>
         </svg>
         <div
```

Comment:
Have you considered using D3 just for it's helper methods (e.g. output paths, curve data etc) so the rendering is consistent i.e. via react?

Reply:
Yeah it's something I'm increasingly considering, especially as D3-generated DOM nodes are proving so hard to test with Jest/Enzyme.

I originally generated the chart with D3 over a year ago because I wanted to have more control over the animations. I'm slowly moving towards using React for more of the rendering, but refactoring it would basically only help with adding extra tests at the moment, so it's not a high priority for me.

Reply:
Ok cool, let's keep it in the back of our heads, definitely agree with the testing aspect - might be a bigger endeavour down the line 😄